### PR TITLE
Draft of engineering practices - layering

### PR DESCRIPTION
**Pull request by MerlinDarkstar**
Not sure if image is in correct place - hopefully it makes sense too :)

### CODE REVIEW

File: content/engineering-practice-layering.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+Title: Engineering Practice Layering
+Date: 2023-03-10
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: @MerlinDarkstar can you change this to 2023-03-15?

File: content/engineering-practice-layering.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+Ideally, The Business Rules and Logic should not depend on any other projects in the application.
+
+---
+
+# Different Layering Scenarios
+
+As mentioned above, there are logically 3 layers which can be placed into 3 or more physical Tiers enabled by following this software architecture design pattern.  These layers can then be 
+organized into four distict scenarios, shown here.
+
+![Architecture Scenarios](/Images/n-tier-architecture.png)
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: @MerlinDarkstar the image files must use the same naming convention as the slug. In other words, please rename your image file to: engineering-practice-layering-1.png

File: content/engineering-practice-layering.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+Category: Posts 
+Tags: code, code-quality, engineering, technical-excellence
+Slug: engineering-practice-layering
+Author: Martin M. Lacey
+Summary: Application Layering - Separation of Concerns Architecture Principle
+
+# What is layering?
+
+Layering is a common construct for identifying discreet areas of interest within an undertaking, so that work can be done in a focused way in each of 
+those areas as well as to highlight the interconnections between each layer.  Typically, each layer has its own specialists that work on the specific
```
Comment (wpschaub) [inline content/engineering-practice-layering.md:12]: @MerlinDarkstar what editor are you using? You have paragraph markers in the middle of sentences.

File: content/engineering-practice-layering.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+technology involved in any given layer, and collaborate with other specialists where the layers interface.
+
+This is most easily understood when you look at the physical world - lets take construction for example.  To create a building, first an architectural drawing
+is created that identifies each aspect of construction as a layered endevor.  Once the design is signed-off, a hole is dug, a frame and rebar is placed within and the concrete
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: endeavor

File: content/engineering-practice-layering.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+is poured into the frame - this is the foundation layer.  Next - layer 2 - the wall frames are built for the first floor, the ceiling, the 2nd floor foundation and its' walls; and so on up one floor in the 
+building at a time.  
+
+After each floor has its wall frames in place, layer 3 begins with electrical and plumbing specialists routing their work through the wall frames and up through each level
+of the building.  The 4th layer of work invovles completing the walls with drywall and plaster, and installing the windows.  This continues to the final, finishing layer where the
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: involves

File: content/engineering-practice-layering.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+is created that identifies each aspect of construction as a layered endevor.  Once the design is signed-off, a hole is dug, a frame and rebar is placed within and the concrete
+is poured into the frame - this is the foundation layer.  Next - layer 2 - the wall frames are built for the first floor, the ceiling, the 2nd floor foundation and its' walls; and so on up one floor in the 
+building at a time.  
+
+After each floor has its wall frames in place, layer 3 begins with electrical and plumbing specialists routing their work through the wall frames and up through each level
+of the building.  The 4th layer of work invovles completing the walls with drywall and plaster, and installing the windows.  This continues to the final, finishing layer where the 
+fixtures and appliances are installed, switches installed and connected, paint applied and carpet or hardwood flooring is installed.
+
+This is no different in software - except the layers are logical, containing only code that address a paticular area of interest.  But because it is code, each layer can be ommited or subverted by the unwarry.  Lets examine this more closely next.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: particular
omitted
unwary

File: content/engineering-practice-layering.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+---
+
+# Why is Layering Important?
+
+In Software, *Layering* is a fundamental design principle enabling us to create lasting code of highest quality - and maintainability & testability.  When we layer our software, we
+are aknowledging the different focal points that we need to be concerned with - and possibly employ different specialists to work on each layer independantly.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: acknowledging
independently

File: content/engineering-practice-layering.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+This is no different in software - except the layers are logical, containing only code that address a paticular area of interest.  But because it is code, each layer can be ommited or subverted by the unwarry.  Lets examine this more closely next.
+
+---
+
+# Why is Layering Important?
+
+In Software, *Layering* is a fundamental design principle enabling us to create lasting code of highest quality - and maintainability & testability.  When we layer our software, we
+are aknowledging the different focal points that we need to be concerned with - and possibly employ different specialists to work on each layer independantly. 
+This create a technology
+stack and loosely coupled interfaces between the layers that promote concistency at the communication channels bewteen layers (re: public interfaces), and independance of each layer itself.  Each layer
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: consistency
between
independence

File: content/engineering-practice-layering.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+can also be tested and verified independantly when we employ additional architectual principles such as Dependency Inversion and Encapsulation.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: independently
architectural

File: content/engineering-practice-layering.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+Like the construction 
+example above, this layering represents technology and purpose boundaries, that, when respected, create a modular and consistent design pattern enabling components to be reused and 
+more easily understood.  Indeed, this understanding leads to abstraction of fundamentable behaviours that can organized the code into hierarchies and promote non-duplication of function.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: fundamental

File: content/engineering-practice-layering.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+When we layer, we are more prepared for evolving the software as independent peices in an Agile workflow and able to embrace CI/CD and microservice architecture more readily.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: pieces

File: content/engineering-practice-layering.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+---
+
+# Separation of Conerns Principle
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: Concerns

File: content/engineering-practice-layering.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+Layering is the result of following the *Separation of Concerns* design principle, which states that software should be separated based on the kind
+of work it performs.  When viewed from this lens, we can clearly see what should be in the User Interface collection of code - the code that deals
+with artifacts used to collect or format data for display, and what code performs and enforces business rules and procedures, versus the code used to store and
+retrieve data from a database or file system. 
+
+---
+
+# What are the Layers in Software Architecture
+ 
+Architectually, software applications should be logically constructed following the separation of concerns principle which identifies the application layers.  By separating core business
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: Architecturally

File: content/engineering-practice-layering.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+---
+
+# What are the Layers in Software Architecture
+ 
+Architectually, software applications should be logically constructed following the separation of concerns principle which identifies the application layers.  By separating core business 
+rules and validation logic from user-interface and back-end (data service) infrastructure.  
+
+- **User Interface** - The layer the end-user interacts with, a page or form or visual element such as tables, buttons, charts and graphs, input fields and layout.
+- **Business Logic** - The layer containing the business rules and logical processes, work-flows, and validation rules.
+- **Data Access** - The layer is concerned with the storage and retrieval of data and other artifacts used by the application to perform its funtion.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: function

File: content/engineering-practice-layering.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+Each of these layers is focused on a separate logical area of concern, uses different tools and has a different pupose.  Each of
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: purpose

File: content/engineering-practice-layering.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+these logical layers should be created in separate physical projects within the application solution, which helps enforce the separation of concerns principle, improves modularity and testability.  
+
+Ideally, The Business Rules and Logic should not depend on any other projects in the application.
+
+---
+
+# Different Layering Scenarios
+
+As mentioned above, there are logically 3 layers which can be placed into 3 or more physical Tiers enabled by following this software architecture design pattern.  These layers can then be 
+organized into four distict scenarios, shown here.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: distinct

File: content/engineering-practice-layering.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+![Architecture Scenarios](/Images/n-tier-architecture.png)
+
+- **Classic** - All logical layers are combined into one, making direct access to the underlaying data source from within the single code layer.  This was a common implementation years ago and can still be found, but easily leads to *Boundary Bleed* and messy speghetti code.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: spaghetti

File: content/engineering-practice-layering.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+
+- **Three-Tier** - This is a further refinement and maturation of the software architecture, uniquely placing the business rules (requirements) and logic (processes) within a distinct layer free from presentation and data access specialized coding needs.
+
+- **N-Tier** - This is the holy-grail, where work-flow is separated from the Presentation layer, and accesses the business services layer which uses the data services layer for storage and retrieval; all layers can be distributed across CPU and machine boundaries to achieve scale and reach.  
+
+---
+
+# What is Boundary Bleed 
+
+Boundary Bleed occurs when you violate the separate of concerns principle and allow non built-in types to be exposed through multple layers.  For example, if you
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: multiple

File: content/engineering-practice-layering.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+create a class in your data service layer, and return it to your business service layer, and that in turn returns that same class to the presentation layer; you've
+passed a class through 2 layers and inadvertantly coupled the presentation layer with the data service layer.  This makes your code more brittle and harder to test, where changes
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: inadvertently

File: content/engineering-practice-layering.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+can have unexpected side-effects.  This is *Boundary Bleed*, and it should be avoided and concidered as an anti-pattern.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: considered

File: content/engineering-practice-layering.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+Tags: code, code-quality, engineering, technical-excellence
+Slug: engineering-practice-layering
+Author: Martin M. Lacey
+Summary: Application Layering - Separation of Concerns Architecture Principle
+
+# What is layering?
+
+Layering is a common construct for identifying discreet areas of interest within an undertaking, so that work can be done in a focused way in each of 
+those areas as well as to highlight the interconnections between each layer.  Typically, each layer has its own specialists that work on the specific
+technology involved in any given layer, and collaborate with other specialists where the layers interface.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: layer and ... no comma

File: content/engineering-practice-layering.md (hunk 4 above)
Comment (wpschaub) [inline content/engineering-practice-layering.md]: rebar are ... not is

File: content/engineering-practice-layering.md [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+This is most easily understood when you look at the physical world - lets take construction for example.  To create a building, first an architectural drawing
+is created that identifies each aspect of construction as a layered endevor.  Once the design is signed-off, a hole is dug, a frame and rebar is placed within and the concrete
+is poured into the frame - this is the foundation layer.  Next - layer 2 - the wall frames are built for the first floor, the ceiling, the 2nd floor foundation and its' walls; and so on up one floor in the 
+building at a time.  
+
+After each floor has its wall frames in place, layer 3 begins with electrical and plumbing specialists routing their work through the wall frames and up through each level
+of the building.  The 4th layer of work invovles completing the walls with drywall and plaster, and installing the windows.  This continues to the final, finishing layer where the 
+fixtures and appliances are installed, switches installed and connected, paint applied and carpet or hardwood flooring is installed.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: paint applied, ... add comma?

File: content/engineering-practice-layering.md (hunk 6 above)
Comment (wpschaub) [inline content/engineering-practice-layering.md]: Let's or Let us, not Lets.

File: content/engineering-practice-layering.md [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+This is no different in software - except the layers are logical, containing only code that address a paticular area of interest.  But because it is code, each layer can be ommited or subverted by the unwarry.  Lets examine this more closely next.
+
+---
+
+# Why is Layering Important?
+
+In Software, *Layering* is a fundamental design principle enabling us to create lasting code of highest quality - and maintainability & testability.  When we layer our software, we
+are aknowledging the different focal points that we need to be concerned with - and possibly employ different specialists to work on each layer independantly. 
+This create a technology
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: This creates ...

File: content/engineering-practice-layering.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+---
+
+# What are the Layers in Software Architecture
+ 
+Architectually, software applications should be logically constructed following the separation of concerns principle which identifies the application layers.  By separating core business 
+rules and validation logic from user-interface and back-end (data service) infrastructure.  
+
+- **User Interface** - The layer the end-user interacts with, a page or form or visual element such as tables, buttons, charts and graphs, input fields and layout.
+- **Business Logic** - The layer containing the business rules and logical processes, work-flows, and validation rules.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: workflows

File: content/engineering-practice-layering.md [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+
+- **Classic** - All logical layers are combined into one, making direct access to the underlaying data source from within the single code layer.  This was a common implementation years ago and can still be found, but easily leads to *Boundary Bleed* and messy speghetti code.
+
+- **Two-Tier** - In this scenario, the data access layer is distinctly separated from all other code, which resides in the main project for the application.  This represents an understanding that data access and the object-relation-mapping (ORM) is a distinct aspect of a software application and requires specialized knowledge.
+
+- **Three-Tier** - This is a further refinement and maturation of the software architecture, uniquely placing the business rules (requirements) and logic (processes) within a distinct layer free from presentation and data access specialized coding needs.
+
+- **N-Tier** - This is the holy-grail, where work-flow is separated from the Presentation layer, and accesses the business services layer which uses the data services layer for storage and retrieval; all layers can be distributed across CPU and machine boundaries to achieve scale and reach.
```
Comment (wpschaub) [inline content/engineering-practice-layering.md]: workflow

File: content/engineering-practice-layering.md (hunk 18 above)
Comment (wpschaub) [inline content/engineering-practice-layering.md]: non-built-in?

File: content/engineering-practice-layering.md (hunk 2 above)
Comment (wpschaub) [inline content/engineering-practice-layering.md]: ](../images/...)

Note the lower case "i" in images. GitHub is case sensitive, which is why your image is not showing on the preview.

**Review by wpschaub — DISMISSED** [review]
@MerlinDarkstar just a few typos and non-showing image due to case sensitive gremlin.

File: content/engineering-practice-layering.md (hunk 2 above)
Comment (MerlinDarkstar) [reply content/engineering-practice-layering.md]: I also added the ../images/  - the leading two dots...

Comment (wpschaub) [reply content/engineering-practice-layering.md]: @MerlinDarkstar sorry it is two dots ../images.

Always start with the _template.md file - it keeps me within the guardrails :)

File: content/engineering-practice-layering.md (hunk 3 above)
Comment (MerlinDarkstar) [reply content/engineering-practice-layering.md:12]: I've removed one or two - doesn't seem to make a difference when viewed.  I am using Visual Studio - have the spell checker turned on too but isn't catching any.

**Review by wpschaub — APPROVED** [review]

**Review by AndreKaminski — APPROVED** [review]
Looks great! I am always very cautious when using a comparison of building software with construction industry. The key difference is that in construction you have to follow a linear process, while things in software development might be done in various orders or in parallel. But I think here it works.

**Pull request merged**